tests/integration/backward_compatible/sql_test.py

```diff
@@ -0,0 +1,590 @@
+        six.assertCountEqual(
+            self,
+            [i for i in range(7, entry_count)],
+            [row.get_object("this") for row in result],
+        )
+
+    def test_execute_with_mismatched_params_when_sql_has_more(self):
+        self._populate_map()
+        result = self.client.sql.execute(
+            "SELECT * FROM %s WHERE __key > ? AND this > ?" % self.map_name, 5
+        )
+
+        with self.assertRaises(HazelcastSqlError):
+            for _ in result:
+                pass
+
+    def test_execute_with_mismatched_params_when_params_has_more(self):
+        self._populate_map()
+        result = self.client.sql.execute("SELECT * FROM %s WHERE this > ?" % self.map_name, 5, 6)
+
+        with self.assertRaises(HazelcastSqlError):
+            for _ in result:
+                pass
+
+    def test_execute_statement(self):
+        entry_count = 12
+        self._populate_map(entry_count, str)
+        statement = SqlStatement("SELECT this FROM %s" % self.map_name)
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(
+            self,
+            [str(i) for i in range(entry_count)],
+            [row.get_object_with_index(0) for row in result],
+        )
+
+    def test_execute_statement_with_params(self):
+        entry_count = 20
+        self._populate_map(entry_count, lambda v: Student(v, v))
+        statement = SqlStatement(
+            "SELECT age FROM %s WHERE height = CAST(? AS REAL)" % self.map_name
+        )
+        statement.add_parameter(13.0)
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(self, [13], [row.get_object("age") for row in result])
+
+    def test_execute_statement_with_mismatched_params_when_sql_has_more(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT * FROM %s WHERE __key > ? AND this > ?" % self.map_name)
+        statement.parameters = [5]
+        result = self.client.sql.execute_statement(statement)
+
+        with self.assertRaises(HazelcastSqlError):
+            for _ in result:
+                pass
+
+    def test_execute_statement_with_mismatched_params_when_params_has_more(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT * FROM %s WHERE this > ?" % self.map_name)
+        statement.parameters = [5, 6]
+        result = self.client.sql.execute_statement(statement)
+
+        with self.assertRaises(HazelcastSqlError):
+            for _ in result:
+                pass
+
+    def test_execute_statement_with_timeout(self):
+        entry_count = 100
+        self._populate_map(entry_count, lambda v: Student(v, v))
+        statement = SqlStatement("SELECT age FROM %s WHERE height < 10" % self.map_name)
+        statement.timeout = 100
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(
+            self, [i for i in range(10)], [row.get_object("age") for row in result]
+        )
```

Comment:
What is this test for? I don't why but a query with timeout does not return an error even if it is set to 0. Only setter for timeout is tested in this test

Reply:
0 is not an erroneous case. 0 means there is no timeout. This test is a valid one because the IMDG SQL engine will timeout your queries (based on the value of this) if your query is a long-running one. There is no deterministic way to get timeout exceptions in queries but if you fill your map with enough items and run somewhat complex query, you might get timeouts.

As for the test, I don't think it is an unnecessary one

Reply:
I think the timeout feature is not working yet. Try to set it to 1 and populate map with 1000000 items, then make a complex query; I don't think it will give an error.

Reply:
From the client's PoV, a timeout error is no different than an error response from the server. I don't think we need a separate test for it. As for this test, I don't see a problem keeping it even if the server is not working (yet). We could add an extra test for time outing a query later. Think of this test as setting timeout to something on client and seeing how to server reacts

Reply:
another point is that this test is testing server behaviour(as you said `Think of this test as setting timeout to something on client and seeing how to server reacts`), therefore it is unnecessary for this reason too. As I said above, the only interesting thing left in this test from client's PoV is timeout's setter which is obviously not the aim of this test.

Reply:
Actually it even cannot test server's behaviour due to the fact that it is hard to get a timeout error(I tried but could not get one)

Reply:
I could even leave this here just to test Python's setter, even if this is totally unused in the server

Reply:
Then, I would add a seperate test(and delete this test) for testing the setter with different inputs, not just with one input like here. Because you cannot test this logic with one input:

```python
@timeout.setter
def timeout(self, timeout):
    if timeout < 0 and timeout != SqlStatement.TIMEOUT_NOT_SET:
        raise ValueError("Timeout must be non-negative or -1, not %s" % timeout)

    self._timeout = timeout
```

Reply:
I said even if totally unused :smile: and it is not unused see `SqlInternalService#execute` in Java

Reply:
You intend to test server then. Why you test server side code?

Reply:
As I said we even cannot test server side code, since it does not throw timeout error anyhow

Reply:
I am not sure why we are debating this much in this simple test but I see it as valuable, so if it is not a BIG problem for you, let's leave it like this